### ISSUE TITLE
Support for CoffeeScript classes syntax

### ISSUE THREAD

**Issue by suprMax** [ractivejs/ractive#1885]
Hi!

I'm using CoffeeScript extensively in my app. It provides it's own classes system (http://coffeescript.org/#classes) with ability to call `super` to invoke parent's class method.

When I tried to do `class View extends Ractive` it works for the most part, but fails once you try using partials or components.

Here's an example from Backbone how they handle extends. It works great and I can extend any Backbone class "natively" instead of calling `Backbone.extend`: https://github.com/jashkenas/backbone/blob/master/backbone.js#L1779

Would it be possible to have similar functionality? It seems that most of the functionality you already have anyway (`this._super()` etc). I think CoffeeScript users such as myself would greatly appreciate if it works out of the box natively!

**Comment by martypdx** [comment]
Does it work to do something like:

```js
function CoffeeRactive(options){
   return Ractive.extend(options);
}
```
and go from there?

**Comment by Rich-Harris** [comment]
I think @suprMax is talking about using standard inheritance semantics. I've wondered about this, in the context of ES6 classes (which are somewhat similar to their CoffeeScript counterparts).

The problem is that the semantics are a bit different. When you call `A = Ractive.extend(...)`, followed by `B = A.extend()`, you're not just creating a constructor whose prototype inherits from the parent; 'registries' (such as `partials` and `components`, as you note, but also `transitions`, `events`, `decorators` and so on) plus `data` and `computed` all extend the corresponding properties on the parent.

More than that, it allows Ractive to do things like only compiling templates once at `extend` time, rather than at instantiation time. So it's a lot more efficient if you're extending a component and only making minor changes (for example, `WarningModal = Modal.extend({ data: { type: 'warning' } })`, or something) - everything 'rolls up' neatly, saving both you and the computer work.

So I don't really know. React recently moved in this direction (preferring `class Foo extends React.Component` over `React.createClass()`, and it's fair to say it's causing some of their users a bit of pain. Going with the grain of the language is generally a good thing, but I think we'd lose a lot of convenience in the process. It would involve quite a wide-ranging rethink, I think.

**Comment by martypdx** [comment]
As long as you're willing to forgo registry rollup (or handle on your own), I'm thinking that the extend options could be treated as the "class":

```js
function RactiveClass(options) {
    var Extended = Ractive.extend( Object.getPrototypeOf(this) );
    return new Extended(options);
}
```
Then that should enable:
```js
class Widget extends RactiveClass {

    template: '#widget'

    oninit () {

    }

    close () {
        // method call for event
    }
}

var ractive = new Widget({ el: 'body' })

**Comment by suprMax** [comment]
@Rich-Harris I think if you precompile templates (for example with browserify transform), then there is no benefit of compiling it on `Ractive.extend`.

There is indeed a problem of merging all the properties of the prototypes (events, partials etc etc). It has been solved by lovely chaps from ChaplinJS, here's an example: https://github.com/chaplinjs/chaplin/blob/master/src/chaplin/lib/utils.coffee#L68

Basically they walk through prototype chain of extended objects to collect all those props what were set in different places.

I think the best way to tackle this would be to allow the user of the framework to decide which approach to use: `Ractive.extend` or `class Blah extends Ractive`. Backbone guys leave you with both options and it's been working great so far.

**Comment by Crisfole** [comment]
Would something like this work?

```coffeescript
class CoffeeRactive
    constructor: () ->
        return Ractive.extend(this)
```

It compiles down to:

```javascript
var CoffeeRactive;

CoffeeRactive = (function() {
  function CoffeeRactive() {
    return Ractive.extend(this);
  }

  return CoffeeRactive;

})();
```

**Comment by martypdx** [comment]
this exposes a couple of bugs in the ractive code. This works for new instances, but not inline components (still looking into that). @Crisfole basically same approach, but allows inclusion of new options on instantiation.

```coffeescript
class RactiveClass
  constructor: (options) ->
    proto = Object.getPrototypeOf(this);
    delete proto.constructor;
    Extended = Ractive.extend( proto )
    new Extended(options)

class View extends RactiveClass
  oninit: ->
    @on 'select', ->
        console.log 'click!'
        
    @observe 'foo', (n, o) ->
        console.log o, 'to', n
        

view = new View
    el: 'main'
    template: '#template'
    data:
        foo: 'bar'
```

**Comment by suprMax** [comment]
Default: http://jsfiddle.net/47p3xLks/

Proposed by @Crisfole: http://jsfiddle.net/47p3xLks/1/ (doesn't work, doesn't try to render)

@martypdx yup works if you pass some of the props as options, but if you don't and pass only data (template and el set on proto - no luck) http://jsfiddle.net/47p3xLks/2/

My sauce:

````coffeescript
class RactiveClass
  constructor: (options) ->
    proto = Object.getPrototypeOf(this)
    delete proto.constructor
    Extended = Ractive.extend(proto)
    new Extended(options)

class View extends RactiveClass
  doStuff: ->

class TestView extends View
  el: '#main'
  template: "Hello {{name}}!"

test = new TestView(data: name: 'world')
console.warn test

````

**Comment by Crisfole** [comment]
Hmm...something's off: http://jsfiddle.net/dnhobx9j/1/

**Comment by martypdx** [comment]
Turns out to not be so straightforward in the inline component case. Need to look into it deeper.

**Comment by wreckah** [comment]
Have the same problem with coffee :+1:

**Comment by codler** [comment]
This is how I have solved it in TypeScript https://gist.github.com/codler/c91bc144a822c3fb0fc2 
#1687 (for 0.7)

**Comment by evs-chris** [comment]
Having recently gone through all of the extend code, I can say that direct classes are not in the cards for Ractive. There's just too much that depends on one pass through the extension phase.
